staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher.go

```diff
@@ -138,6 +167,82 @@ func (w *watcher) createWatchChan(ctx context.Context, key string, rev int64, re
 	return wc
 }
 
+// isInitialEventsEndBookmarkRequired since there is no way to directly set
+// opts.ProgressNotify from the API and the etcd3 impl doesn't support
+// notification for external clients we simply return initialEventsEndBookmarkRequired
+// to only send the bookmark event after the initial list call.
+//
+// see: https://github.com/kubernetes/kubernetes/issues/120348
+func (w *watcher) isInitialEventsEndBookmarkRequired(opts storage.ListOptions) (bool, error) {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.WatchList) {
+		return false, nil
+	}
+	if opts.SendInitialEvents == nil || !*opts.SendInitialEvents || !opts.Predicate.AllowWatchBookmarks {
+		return false, nil
+	}
+	if err := w.validateRequiredFieldsForWatchListFeature(); err != nil {
+		return false, err
+	}
+	return true, nil
+}
+
+// getStartWatchResourceVersion returns a ResourceVersion
+// the watch will be started from.
+// Depending on the input parameters the semantics of the returned ResourceVersion are:
+//   - start at Exact (return currentRV)
+//   - start at Most Recent (return an RV from etcd)
+func (w *watcher) getStartWatchResourceVersion(ctx context.Context, currentRV int64, opts storage.ListOptions) (int64, error) {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.WatchList) {
+		return currentRV, nil
+	}
+	if opts.SendInitialEvents == nil || *opts.SendInitialEvents {
+		return currentRV, nil
+	}
+	if err := w.validateRequiredFieldsForWatchListFeature(); err != nil {
+		return 0, err
+	}
+	currentStorageRV, err := storage.GetCurrentResourceVersionFromStorage(ctx, w.storage, w.newListFunc, w.resourcePrefix, w.objectType)
+	if err != nil {
+		return 0, fmt.Errorf("failed to get the current resource version from the storage: %w", err)
+	}
+	// currentStorageRV is taken from resp.Header.Revision (int64)
+	// and cast to uint64, so it is safe to do reverse
+	// at some point we should unify the interface but that
+	// would require changing  Versioner.UpdateList
+	return int64(currentStorageRV), nil
```

Comment:
Hmm - this seems wrong to me - shouldn't we use that only if the passed RV is 0 or unset?

Reply:
ups, good catch, thanks!

Reply:
Added below - resolving.